### ISSUE TITLE
[BUILD] OpenVDB in Python on Win10

### ISSUE THREAD

**Issue by petulka8s** [AcademySoftwareFoundation/openvdb#946]
**Operating System:** (e.g. Windows 2019, macOS-10.15, Linux CentOS 7.4)
**Version / Commit SHA:** (e.g VDB 7.2, SHA: 1a2b3c4d5e...)
**CMake Version:** (e.g. 3.18)
**Compiler:** (e.g. gcc 10, msvc 2019)

Hi,
is there possibility to use **OpenVDB in Python on Win10?**
Because when I install openvdb package and import library

**>>> import pyopenvdb**
or
**>>> import openvdb**

 I obtain following error:
**_ImportError: LD_LIBRARY_PATH environment variable not set. The pyopenvdb package currently only work on Linux x86_64 systems._**

Is there a way how can I use openvdb library on Win 10 (exclude c++) or how can I create openvdb volume model in Python by using your library?

Thank you very much for your advice.
Have a nice day.

**Comment by Idclip** [comment]
Hello @petulka8s - how are you currently installing the openvdb python lib?

**Comment by petulka8s** [comment]
Hi,

I am using openvdb 8.0.0 and there is option OPENVDB_BUILD_PYTHON MODULE which I checked and set paths to all libraries: Blosc, Boost, IlmBase, OpenEXR, tbb, zlib.

I can successfully build it. It generates file **pyopenvdb.pyd**.
But when I want to import it to python by using: **import pyopenvdb**
I get following: _Exception has occurred: ImportError. DLL load failed while importing pyopenvdb_

So I tried to run RUN_TESTS and there I am getting setlocal error.
I am not sure what could be wrong.

**Comment by nranthony** [comment]
Hi @petulka8s,
I just wanted to join and let you know that I'm having exactly the same troubles.  I tried to use the CMake GUI and perform the same steps, but went down the rabbit hole of providing paths and libs for a long list of different packages.  In the end I added `vcpkg install boost-python:x64-windows` and then added the python module to the cmd line cmake:
`cmake -DCMAKE_TOOLCHAIN_FILE=C:\dev\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -A x64 -DOPENVDB_BUILD_PYTHON_MODULE=ON ..`
This built a solution, which I then built, and running VS2019 as admin, then installed.  I now have the OpenVDB folder in Program Files.

Alas, I get the same error on importing pyopenvdb.

Following on from this post on .pyd files: https://stackoverflow.com/a/60208567 I assume the PyInit must be looking in the wrong places.

Are there a set of required paths that are needed for importing pyopenvdb.pyd?

**Comment by nranthony** [comment]
Ok, so I should convey my progress.  Not tested to any degree, but promising.

- I found that I could check dependencies with this cool tool:  https://github.com/lucasg/Dependencies
- I first found that the openvdb.dll was missing, so I copied that from the Release folder to the python\Release folder.
- I then got an error that python39.dll would not work with the version being used: 3.8
- As anaconda does not support 3.9 yet without some work I installed 3.9 stand alone into C:\Python3_9_1, and pip installed spyder-kernels once cd'ing that folder: `python -m pip install spyder-kernels` (to use the pip that's in that folder).
- I then got the same DLL error, but on retrying the Dependencies app a few times I found that I needed `snappy.dll`, `zstd.dll`, `lz4.dll`, and `blosc.dll`
- After copying all those into the python\Release folder I could then get Spyder to import pyopenvdb :)
- Note, I had to set the Preferences -> Python interpreter to use the defined Python3_9_1 python.exe, and restarted the kernel, together with adding the below lines prior to the import.
`import sys
sys.path.append(r'C:\dev\openvdb\build\openvdb\openvdb\python\Release')`

I'd love to hear if there's a more correct way to do this (which I'm sure there is), and/or if there's a way to use python38.dll in building the entire openvdb package from the start point.  Is there a way to set the boost-python version maybe?

**Comment by Idclip** [comment]
> In the end I added vcpkg install boost-python:x64-windows and then added the python module to the cmd line cmake:
cmake -DCMAKE_TOOLCHAIN_FILE=C:\dev\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -A x64 -DOPENVDB_BUILD_PYTHON_MODULE=ON ..

This is correct. The [dependency page](https://www.openvdb.org/documentation/doxygen/dependencies.html#depComponents) highlights this requirement and [the readme](https://github.com/AcademySoftwareFoundation/openvdb) provides an example usage with vcpkg. 

> I then got an error that python39.dll would not work with the version being used: 3.8

Your boost-python version needs to be compatible with the version of python you use.

> Are there a set of required paths that are needed for importing pyopenvdb.pyd?

We do not copy dep DLLs to the python installation folder, instead we expect you to have your environment correctly configured. This changes based on how you have installed your own dependencies so is your responsibility. If you are using vcpkg I suggest reading the vcpkg docs (specifically look at `vcpkg.exe help env`) and [the windows DLL search documentation](https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#search-order-for-desktop-applications). You can also modify your system PATH directly or do this in python before importing (from py3.8 you should use https://docs.python.org/3/library/os.html#os.add_dll_directory). 

I am not clear on what the Windows paradigm is with regards to copying DLLs all over the place; where does Windows draw the line with expecting your PATH to be correct vs trying to copy everything everywhere. It seems like either may be valid, and in this case we require the prior.

**Comment by nranthony** [comment]
Hi @Idclip, thanks for the speedy reply.

Thanks for pointing me to a number of really useful docs.  I find it's a minefield if you don't know where to look.  On that note, and general advice to read docs before reaching out, I've tried to find out how to specify the python version of boost-python, but stuck again.  Is that something you can point me too, or a better question for stack exchange?

Thanks for your time.  This is really really helpful.

**Comment by petulka8s** [comment]
Hi,

I am using Cmake GUI instead of vcpkg. I set all paths correct. 

![cmake_openvdb](https://user-images.githubusercontent.com/35227032/107156271-b09f7200-697d-11eb-93c8-45402b5759a4.png)

I am able to generate it and build solution.
The resulting folder has follow structure:

![folder_structure](https://user-images.githubusercontent.com/35227032/107156321-fb20ee80-697d-11eb-9bdd-898df06f368e.png)

I took all possible .dll files which I need (Blosc, Boost, IlmBase, OpenEXR, OpenVDB, tbb, zlib) and put it to same folder as my script where I want to import pyopenvdb.pyd. It didnt work so I put everything to python path. But probably I miss some important library like snappy.dll. Python version is 3.8 as same as boost-python. I use version Boost1.72.

**Comment by petulka8s** [comment]
@nranthony do you have any success?

**Comment by petulka8s** [comment]
Hi guys,

I am using vcpkg install as you mentioned.

But the problem is with the newest versions of libraries. The boost 1.75 library is not compatile with project.

How can I install older version of libaries? For example, vcpkg install boost 1.72?

Thanks for you reply.

**Comment by nranthony** [comment]
Alas, I have not found a way to deal with anything but the boost-python version that gets installed.  I just installed python 39 and got the python vdb test script to work there.

I'm going to start a post on stack exchange as this is not a vdb issue per se, and more or a versions and dependencies thing.  I did look into the cmake files, and there's a FindBoost.cmake file in the cmake folders that details:

`Note that Boost Python components require a Python version suffix
(Boost 1.67 and later), e.g. ``python36`` or ``python27`` for the
versions built against Python 3.6 and 2.7, respectively.  This also
applies to additional components using Python including
``mpi_python`` and ``numpy``.  Earlier Boost releases may use
distribution-specific suffixes such as ``2``, ``3`` or ``2.7``.
These may also be used as suffixes, but note that they are not
portable.`

and then goes on to provide an example:
`Example to find Boost Python 3.6 libraries and use imported targets::

  find_package(Boost 1.67 REQUIRED COMPONENTS
               python36 numpy36)
  add_executable(foo foo.cc)
  target_link_libraries(foo Boost::python36 Boost::numpy36)`

which means that it's an option at some point, but I can't seem to find the flag/parameter that would be applicable, and/or where to define or edit.

**Comment by nranthony** [comment]
See https://stackoverflow.com/questions/66110606/how-to-define-boost-python-version-when-using-vcpkg in parallel to this.

This issue should really be marked closed given it works if you use python 3.9

Neil

**Comment by Idclip** [comment]
> But the problem is with the newest versions of libraries. The boost 1.75 library is not compatile with project.

Hello @petulka8s

I was able to get the python module built with no issues with boost 1.75 and python 3.9 on Windows 10 from scratch. I installed the latest vcpkg, git and cmake (3.19). This is my workflow:

Install deps:
```
vcpkg install openexr blosc tbb boost --triplet=x64-windows
```

Build VDB:
```
mkdir build
cd build
cmake -DOPENVDB_BUILD_PYTHON_MODULE=ON -DCMAKE_TOOLCHAIN_FILE="<PATH_TO_VCPKG>\scripts\buildsystems\vcpkg.cmake" -DVCPKG_TARGET_TRIPLET=x64-windows -A x64 ..
cmake --build . --parallel 4 --config Release --target install
```

Finally, to run, navigate to the location of `pyopenvdb.pyd`:
```
> python
>>> import os
>>> os.add_dll_directory("<PATH_TO_VCPKG>/installed/x64-windows/bin")  # e.g. "C:/dev/vcpkg/installed/x64-windows/bin"
>>> os.add_dll_directory("<PATH_TO_VDB_INSTALL>/bin")                             # e.g. "C:/Program Files/OpenVDB/bin
>>> import pyopenvdb
```

Note that there are some important changes with python 3.8/3.9:
https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew

> DLL dependencies for extension modules and DLLs loaded with ctypes on Windows are now resolved more securely. Only the system paths, the directory containing the DLL or PYD file, and directories added with add_dll_directory() are searched for load-time dependencies. Specifically, PATH and the current working directory are no longer used, and modifications to these will no longer have any effect on normal DLL resolution. If your application relies on these mechanisms, you should check for add_dll_directory() and if it exists, use it to add your DLLs directory while loading your library.

**Comment by petulka8s** [comment]
Thanks for your advices @Idclip.
It helped me a lot.
I set the paths using **os.add_dll_directory(path_to_folder)** and now its working!

Once again, thank you!!!

**Comment by Idclip** [comment]
No problem, I will close this! 

@nranthony please feel free to open a new issue regarding boost-python if you suspect it to be an problem with OpenVDB's CMake. I'm a bit unclear at the moment exactly where your issue sits.

**Comment by PPPoseidon** [comment]
> > But the problem is with the newest versions of libraries. The boost 1.75 library is not compatile with project.
> 
> Hello @petulka8s
> 
> I was able to get the python module built with no issues with boost 1.75 and python 3.9 on Windows 10 from scratch. I installed the latest vcpkg, git and cmake (3.19). This is my workflow:
> 
> Install deps:
> 
> ```
> vcpkg install openexr blosc tbb boost --triplet=x64-windows
> ```
> 
> Build VDB:
> 
> ```
> mkdir build
> cd build
> cmake -DOPENVDB_BUILD_PYTHON_MODULE=ON -DCMAKE_TOOLCHAIN_FILE="<PATH_TO_VCPKG>\scripts\buildsystems\vcpkg.cmake" -DVCPKG_TARGET_TRIPLET=x64-windows -A x64 ..
> cmake --build . --parallel 4 --config Release --target install
> ```
> 
> Finally, to run, navigate to the location of `pyopenvdb.pyd`:
> 
> ```
> > python
> >>> import os
> >>> os.add_dll_directory("<PATH_TO_VCPKG>/installed/x64-windows/bin")  # e.g. "C:/dev/vcpkg/installed/x64-windows/bin"
> >>> os.add_dll_directory("<PATH_TO_VDB_INSTALL>/bin")                             # e.g. "C:/Program Files/OpenVDB/bin
> >>> import pyopenvdb
> ```
> 
> edit: Alternatively, you do not need to navigate to the location of `pyopenvdb.pyd`, just set your `PYTHONPATH` environment variable to point there.
> 
> Note that there are some important changes with python 3.8/3.9: https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
> 
> > DLL dependencies for extension modules and DLLs loaded with ctypes on Windows are now resolved more securely. Only the system paths, the directory containing the DLL or PYD file, and directories added with add_dll_directory() are searched for load-time dependencies. Specifically, PATH and the current working directory are no longer used, and modifications to these will no longer have any effect on normal DLL resolution. If your application relies on these mechanisms, you should check for add_dll_directory() and if it exists, use it to add your DLLs directory while loading your library.

thanks!!!!

**Comment by keli95566** [comment]
Build with NumPy option will throw the following errors though :/ Anyone knows a fix? (On windows) 

-- Could NOT find Python (missing: Python_LIBRARIES Python_NumPy_INCLUDE_DIRS Development.Module NumPy) (found version "3.10.5")
-- Found Python::Module: ........) (found suitable version "3.10.5", minimum required is "2.7")
CMake Error at openvdb/openvdb/python/CMakeLists.txt:66 (message):
  Could NOT find Python::NumPy (Required is at least version "1.17.0")
Call Stack (most recent call first):
  openvdb/openvdb/python/CMakeLists.txt:129 (openvdb_check_python_version)
